In [133]:
# Packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [134]:
play_df = pd.read_csv('googleplaystore.csv')
review_df = pd.read_csv('googleplaystore_user_reviews.csv')

## Notes:



Website for dataset:
https://www.kaggle.com/lava18/google-play-store-apps

# EDA

## Accessing

### Play_df review

In [135]:
# Looking at data
play_df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [136]:
# Dimensions
play_df.shape

(10841, 13)

Currently, data types are mostly objects. Need to update this

In [137]:
play_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
App               10841 non-null object
Category          10841 non-null object
Rating            9367 non-null float64
Reviews           10841 non-null object
Size              10841 non-null object
Installs          10841 non-null object
Type              10840 non-null object
Price             10841 non-null object
Content Rating    10840 non-null object
Genres            10841 non-null object
Last Updated      10841 non-null object
Current Ver       10833 non-null object
Android Ver       10838 non-null object
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [138]:
play_df.describe()

Rating
count  9367.000000
mean      4.193338
std       0.537431
min       1.000000
25%       4.000000
50%       4.300000
75%       4.500000
max      19.000000

Might see if we can convert all the sizes to MB

In [139]:
# Displaying sizes of all apps
play_df.Size.value_counts().head()

Varies with device    1695
11M                    198
12M                    196
14M                    194
13M                    191
Name: Size, dtype: int64

In [140]:
play_df.Size.value_counts().tail()

619k    1
728k    1
872k    1
569k    1
658k    1
Name: Size, dtype: int64

Looking at different series. Might look at different caterogies of apps.

In [141]:
play_df.nunique()

App               9660
Category            34
Rating              40
Reviews           6002
Size               462
Installs            22
Type                 3
Price               93
Content Rating       6
Genres             120
Last Updated      1378
Current Ver       2832
Android Ver         33
dtype: int64

In [142]:
play_df.Type.value_counts()

Free    10039
Paid      800
0           1
Name: Type, dtype: int64

In [143]:
# Need to fix this
play_df.loc[play_df['Type'] == '0']

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

In [144]:
# Looking at different Content Ratings
play_df['Content Rating'].value_counts()

Everyone           8714
Teen               1208
Mature 17+          499
Everyone 10+        414
Adults only 18+       3
Unrated               2
Name: Content Rating, dtype: int64

In [145]:
# Filtering for Unrated
play_df.loc[play_df['Content Rating'] == 'Unrated']

App Category  Rating Reviews  Size Installs  Type  \
7312     Best CG Photography   FAMILY     NaN       1  2.5M     500+  Free   
8266  DC Universe Online Map    TOOLS     4.1    1186  6.4M  50,000+  Free   

     Price Content Rating         Genres       Last Updated Current Ver  \
7312     0        Unrated  Entertainment      June 24, 2015         5.2   
8266     0        Unrated          Tools  February 27, 2012         1.3   

       Android Ver  
7312    3.0 and up  
8266  2.3.3 and up

In [146]:
# Duplicate values
play_df.duplicated().sum()

483

In [147]:
# Seeing list of duplicates
play_df[play_df.duplicated(keep=False)].sort_values('App').head()

App            Category  Rating Reviews  Size  \
1393       10 Best Foods for You  HEALTH_AND_FITNESS     4.0    2490  3.8M   
1407       10 Best Foods for You  HEALTH_AND_FITNESS     4.0    2490  3.8M   
2543  1800 Contacts - Lens Store             MEDICAL     4.7   23160   26M   
2322  1800 Contacts - Lens Store             MEDICAL     4.7   23160   26M   
2256  2017 EMRA Antibiotic Guide             MEDICAL     4.4      12  3.8M   

        Installs  Type   Price Content Rating            Genres  \
1393    500,000+  Free       0   Everyone 10+  Health & Fitness   
1407    500,000+  Free       0   Everyone 10+  Health & Fitness   
2543  1,000,000+  Free       0       Everyone           Medical   
2322  1,000,000+  Free       0       Everyone           Medical   
2256      1,000+  Paid  $16.99       Everyone           Medical   

           Last Updated Current Ver   Android Ver  
1393  February 17, 2017         1.9  2.3.3 and up  
1407  February 17, 2017         1.9  2.3.3 and up  
2543      July 27, 2018       7.4.1    5.0 and up  
2322      July 27, 2018       7.4.1    5.0 and up  
2256   January 27, 2017       1.0.5  4.0.3 and up

In [148]:
# Null values
play_df.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

Handling nulls

In [149]:
# For blank ratings, delete files
play_df.loc[play_df['Rating'].isnull()].head()

App        Category  Rating Reviews  \
23                Mcqueen Coloring pages  ART_AND_DESIGN     NaN      61   
113            Wrinkles and rejuvenation          BEAUTY     NaN     182   
123               Manicure - nail design          BEAUTY     NaN     119   
126         Skin Care and Natural Beauty          BEAUTY     NaN     654   
129  Secrets of beauty, youth and health          BEAUTY     NaN      77   

     Size  Installs  Type Price Content Rating  \
23   7.0M  100,000+  Free     0       Everyone   
113  5.7M  100,000+  Free     0   Everyone 10+   
123  3.7M   50,000+  Free     0       Everyone   
126  7.4M  100,000+  Free     0           Teen   
129  2.9M   10,000+  Free     0     Mature 17+   

                              Genres        Last Updated Current Ver  \
23   Art & Design;Action & Adventure       March 7, 2018       1.0.0   
113                           Beauty  September 20, 2017         8.0   
123                           Beauty       July 23, 2018         1.3   
126                           Beauty       July 17, 2018        1.15   
129                           Beauty      August 8, 2017         2.0   

    Android Ver  
23   4.1 and up  
113  3.0 and up  
123  4.1 and up  
126  4.1 and up  
129  2.3 and up

In [150]:
# Just going to delete this
play_df.loc[play_df['Content Rating'].isnull()]

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

### Review_df

In [151]:
review_df.head(10)

App                                  Translated_Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   
5  10 Best Foods for You                                           Best way   
6  10 Best Foods for You                                            Amazing   
7  10 Best Foods for You                                                NaN   
8  10 Best Foods for You                               Looking forward app,   
9  10 Best Foods for You              It helpful site ! It help foods get !   

  Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0  Positive                1.00                0.533333  
1  Positive                0.25                0.288462  
2       NaN                 NaN                     NaN  
3  Positive                0.40                0.875000  
4  Positive                1.00                0.300000  
5  Positive                1.00                0.300000  
6  Positive                0.60                0.900000  
7       NaN                 NaN                     NaN  
8   Neutral                0.00                0.000000  
9   Neutral                0.00                0.000000

Definition of Sentiment Polarity and Sentiment Subjectivity.

Might relate to NLP (Natural Language Processing)

Sentiment Polarity: It simply means emotions expressed in a sentence.

Sentiment Subjectivity: Subjective sentence expresses some personal feelings, views, or beliefs.

Sources<br>
https://www.quora.com/What-is-polarity-and-subjectivity-in-sentiment-analysis

https://monkeylearn.com/sentiment-analysis/

https://www.kdnuggets.com/2018/08/emotion-sentiment-analysis-practitioners-guide-nlp-5.html

If interested in learning more about NLP <br>
https://activewizards.com/blog/comparison-of-python-nlp-libraries/

In [152]:
review_df.shape

(64295, 5)

In [153]:
review_df.describe()

Sentiment_Polarity  Sentiment_Subjectivity
count        37432.000000            37432.000000
mean             0.182146                0.492704
std              0.351301                0.259949
min             -1.000000                0.000000
25%              0.000000                0.357143
50%              0.150000                0.514286
75%              0.400000                0.650000
max              1.000000                1.000000

In [154]:
review_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
App                       64295 non-null object
Translated_Review         37427 non-null object
Sentiment                 37432 non-null object
Sentiment_Polarity        37432 non-null float64
Sentiment_Subjectivity    37432 non-null float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


In [155]:
# Duplicate values
review_df.duplicated().sum()

33616

In [156]:
# Seeing list of duplicates
review_df[review_df.duplicated(keep=False)].sort_values('Translated_Review').head(20)

App  \
16025                                     Block Puzzle   
15925                                     Block Puzzle   
15885                                     Block Puzzle   
20401                     CNN Breaking US & World News   
20361                     CNN Breaking US & World News   
20441                     CNN Breaking US & World News   
12305                      Banfield Pet Health Tracker   
12265                      Banfield Pet Health Tracker   
22173                     Calorie Counter - MyNetDiary   
22273                     Calorie Counter - MyNetDiary   
21756                         Calorie Counter - Macros   
21856                         Calorie Counter - Macros   
41229  Evernote – Organizer, Planner for Notes & Memos   
41189  Evernote – Organizer, Planner for Notes & Memos   
57909                   Gyft - Mobile Gift Card Wallet   
57949                   Gyft - Mobile Gift Card Wallet   
57989                   Gyft - Mobile Gift Card Wallet   
42371                    FOX NOW - On Demand & Live TV   
42331                    FOX NOW - On Demand & Live TV   
35061                    Docs To Go™ Free Office Suite   

                                       Translated_Review Sentiment  \
16025  "Classic" allows untimed play; game holds posi...  Positive   
15925  "Classic" allows untimed play; game holds posi...  Positive   
15885  "Classic" allows untimed play; game holds posi...  Positive   
20401  "Real" news MAJORITY public wants hear. While ...  Positive   
20361  "Real" news MAJORITY public wants hear. While ...  Positive   
20441  "Real" news MAJORITY public wants hear. While ...  Positive   
12305  "View Medical History" Status Center work. I u...  Negative   
12265  "View Medical History" Status Center work. I u...  Negative   
22173  $40 year want features. Looks though could pay...  Positive   
22273  $40 year want features. Looks though could pay...  Positive   
21756  'search online' simply work, freezes time I tr...  Negative   
21856  'search online' simply work, freezes time I tr...  Negative   
41229  (Add Theme Feature) I wish Black theme. I like...  Negative   
41189  (Add Theme Feature) I wish Black theme. I like...  Negative   
57909  ***Update 11/19/17 Broken problem ------------...  Positive   
57949  ***Update 11/19/17 Broken problem ------------...  Positive   
57989  ***Update 11/19/17 Broken problem ------------...  Positive   
42371  **Update** Yet another glitch app. Can't cast?...  Positive   
42331  **Update** Yet another glitch app. Can't cast?...  Positive   
35061  *TRIED THE TOP 5 DOCUMENT APPS, including this...  Positive   

       Sentiment_Polarity  Sentiment_Subjectivity  
16025            0.033333                0.355556  
15925            0.033333                0.355556  
15885            0.033333                0.355556  
20401            0.130527                0.470408  
20361            0.130527                0.470408  
20441            0.130527                0.470408  
12305           -0.075000                0.066667  
12265           -0.075000                0.066667  
22173            0.388889                0.616667  
22273            0.388889                0.616667  
21756           -0.033333                0.426190  
21856           -0.033333                0.426190  
41229           -0.047222                0.248611  
41189           -0.047222                0.248611  
57909            0.089394                0.454924  
57949            0.089394                0.454924  
57989            0.089394                0.454924  
42371            0.137245                0.407483  
42331            0.137245                0.407483  
35061            0.112879                0.583636

In [157]:
# Null values
review_df.isnull().sum()

App                           0
Translated_Review         26868
Sentiment                 26863
Sentiment_Polarity        26863
Sentiment_Subjectivity    26863
dtype: int64

In [158]:
review_df[review_df.Translated_Review.isnull()].head()
# review_df[(review_df.Translated_Review.isnull()) and (review_df.Translated_Review != isnull()) ]

App Translated_Review Sentiment  Sentiment_Polarity  \
2    10 Best Foods for You               NaN       NaN                 NaN   
7    10 Best Foods for You               NaN       NaN                 NaN   
15   10 Best Foods for You               NaN       NaN                 NaN   
102  10 Best Foods for You               NaN       NaN                 NaN   
107  10 Best Foods for You               NaN       NaN                 NaN   

     Sentiment_Subjectivity  
2                       NaN  
7                       NaN  
15                      NaN  
102                     NaN  
107                     NaN

Things to look into
look at what sentiment polarity and subjectivity is about

### Things to look into
> <b>play_df</b>
<br><br>
- Take out duplicates
- Remove apps without ratings
- convert to integer: size, price, installs
- For size, convert to MB
- Change "Last update" into a date time
- Delete row 10472
- Remove: Genres, Current Version, Android Ver

> <b>review_df</b><br>
- delete dups
- delete NaNs that are in Translated_Review, Sentiment, Sentiment_Polarity, Sentiment_Subjectivity
- for remaining Nans, p
- get mean polarity and subjectivity for each app and add to play_df

## Append

### play_df

Removing duplicates

In [159]:
# Dropping dups
play_df.drop_duplicates(inplace=True)

In [160]:
# Duplicate values
play_df.duplicated().sum()

0

Ratings that are nulls will have 'N/A'

In [161]:
# Replacing NaNs
play_df.Rating.dropna(inplace=True)

In [162]:
# Count of NaNs, post-replace
play_df.Rating.isna().sum()

0

convert to integer: size, price, installs

<b>- Size: </b> <br>
Converting all data sizes to one, which will be megabytes

Using `series.str` method, accessing the measurement and data size types


In [180]:
# : M or K (last value of string)
play_df['Size'].str[-1].head() 

0    M
1    M
2    M
3    M
4    M
Name: Size, dtype: object

In [184]:
# the value of the size
play_df['Size'].str[:-1].head() 

0     19
1     14
2    8.7
3     25
4    2.8
Name: Size, dtype: object

Checking to see the types of data. Other than kilobytes and megabytes, there seems to be an additional 1500 apps that are 'e'

In [182]:
play_df['Size'].str[-1].value_counts() 

M    8516
e    1526
k     315
+       1
Name: Size, dtype: int64

In [191]:
# Will changed this to np.nan
play_df[play_df['Size'].str[-1] == 'e'].head(3)

App           Category  Rating Reviews  \
37             Floor Plan Creator     ART_AND_DESIGN     4.1   36639   
42     Textgram - write on photos     ART_AND_DESIGN     4.4  295221   
52  Used Cars and Trucks for Sale  AUTO_AND_VEHICLES     4.6   17057   

                  Size     Installs  Type Price Content Rating  \
37  Varies with device   5,000,000+  Free     0       Everyone   
42  Varies with device  10,000,000+  Free     0       Everyone   
52  Varies with device   1,000,000+  Free     0       Everyone   

             Genres   Last Updated         Current Ver         Android Ver  
37     Art & Design  July 14, 2018  Varies with device        2.3.3 and up  
42     Art & Design  July 30, 2018  Varies with device  Varies with device  
52  Auto & Vehicles  July 30, 2018  Varies with device  Varies with device

In [192]:
# This will be deleted
play_df[play_df['Size'].str[-1] == '+'].head(3)

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

In [132]:
play_df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  KorM  
0  4.0.3 and up   19M  
1  4.0.3 and up   14M  
2  4.0.3 and up  8.7M  
3    4.2 and up   25M  
4    4.4 and up  2.8M

- <b> Price </b> <br>

- Installs

### review_df

# Analysis and Visuals

Look at:
- Best free and paid apps
- Best in catergory
- Best in Content rating
- Best small and large sized apps